go/pools/fast_pool.go

```diff
@@ -0,0 +1,604 @@
+// till the necessary number of resources are returned
+// to the pool.
+//
+// A SetCapacity of 0 is equivalent to closing the pool.
+func (p *FastPool) SetCapacity(capacity int, block bool) error {
+	p.Lock()
+
+	if p.state.Closed {
+		p.Unlock()
+		return ErrClosed
+	}
+
+	if capacity < 0 || capacity > cap(p.pool) {
+		p.Unlock()
+		return fmt.Errorf("capacity %d is out of range", capacity)
+	}
+
+	if capacity > 0 {
+		minActive := p.state.MinActive
+		if capacity < minActive {
+			p.Unlock()
+			return fmt.Errorf("minActive %v would now be higher than capacity %v", minActive, capacity)
+		}
+	}
+
+	if capacity == 0 {
+		p.state.Closed = true
+	}
+
+	isGrowing := capacity > p.state.Capacity
+	p.state.Capacity = capacity
+
+	if isGrowing {
+		p.state.Draining = false
+		p.Unlock()
+		return nil
+	}
+
+	p.state.Draining = true
+	p.Unlock()
+
+	if block {
+		p.shrink()
+	} else {
+		go p.shrink()
+	}
+
+	return nil
+}
+
+// shrink active resources until capacity it is not above the set capacity.
+func (p *FastPool) shrink() {
+	for {
+		p.Lock()
+		remaining := p.active() - p.state.Capacity
+		if remaining <= 0 {
+			p.state.Draining = false
+			if p.state.Capacity == 0 {
+				close(p.pool)
+				p.state.Closed = true
+			}
+			p.Unlock()
+			return
+		}
+		p.Unlock()
+
+		// We can't remove InUse resources, so only target the pool.
+		// Collect the InUse resources lazily when they're returned.
+		select {
+		case wrapper := <-p.pool:
+			wrapper.resource.Close()
+			wrapper.resource = nil
+
+			p.Lock()
+			p.state.InPool--
+			p.Unlock()
+
+		case <-time.After(time.Second):
+			// Someone could have pulled from the pool just before
+			// we started waiting. Let's check the pool status again.
+		}
+	}
+}
+
+// SetIdleTimeout sets the idle timeout for resources. The timeout is
+// checked at the 10th of the period of the timeout.
+func (p *FastPool) SetIdleTimeout(idleTimeout time.Duration) {
+	p.Lock()
+	p.state.IdleTimeout = idleTimeout
+	p.Unlock()
+}
+
+// closeIdleResources scans the pool for idle resources
+// and closes them. It is meant to run as a goroutine and will
+// return when the pool is closed.
+func (p *FastPool) closeIdleResources() {
+	for !p.IsClosed() {
+
+		timeout := p.State().IdleTimeout
+
+		if timeout == 0 {
+			// Wait for an updated idleTimeout.
+			time.Sleep(time.Second)
+			continue
+		}
+
+		time.Sleep(timeout / 10)
+
+		p.Lock()
+		remaining := p.state.InPool
+		minActive := p.state.MinActive
+		p.Unlock()
+
+		for scanning := true; scanning && remaining > 0; remaining-- {
+			p.Lock()
+			active := p.active()
+			closed := p.state.Closed
+			p.Unlock()
+
+			if closed {
+				return
+			}
+
+			if active <= minActive {
+				break
+			}
+
+			select {
+			case wrapper, ok := <-p.pool:
+				if !ok {
+					return
+				}
+
+				deadline := wrapper.timeUsed.Add(timeout)
+				if time.Now().After(deadline) {
+					p.Lock()
+					p.state.IdleClosed++
+					p.state.InPool--
+					p.Unlock()
```

Comment:
I see a lot of small scattered blocks of code wrapped in a lock.

A useful trick is to create a small helper function like the following:

```go
func (p *FastPool) withLock(f func()) {
  p.Lock()
  f()
  p.Unlock()
}
```

Then call it like so:

```go
p.withLock(func() {
  p.state.IdleClosed++
  p.state.InPool++
})
```

This has the advantage that it guarantees lock/unlock calls are paired, so it's impossible to accidentally miss an unlock. Your call.

Reply:
That looks super handy. I will implement.